framework/include/utils/CompileTimeDerivatives.h

```diff
@@ -101,12 +135,54 @@ class CTUnary : public CTBase
 public:
   CTUnary(T arg) : _arg(arg) {}
 
+  template <typename Self>
+  std::string printParens(const Self *, const std::string & op) const
+  {
+    std::string out = op;
+    if constexpr (T::precedence() > Self::precedence())
+      out += "(" + _arg.print() + ")";
+    else
+      out += _arg.print();
+
+    return out;
+  }
+
   typedef typename T::O O;
 
 protected:
   const T _arg;
 };
 
+/**
+ * Unary minus
+ */
+template <typename T>
+class CTUnaryMinus : public CTUnary<T>
+{
+public:
+  CTUnaryMinus(T arg) : CTUnary<T>(arg) {}
+  using typename CTUnary<T>::O;
+
+  O operator()() const { return -_arg(); }
+  std::string print() const { return this->printParens(this, "-"); }
+  constexpr static int precedence() { return 3; }
```

Comment:
It id the same precedence that the corresponding C++ operators have. Anything else would be ... wrong. I'll put in a comment to that effect.

Reply:
By the way, the precedence and associative property is only needed for stringification of expressions. This is useful when you want to check a derivative. Of course it better be right, otherwise parenthesis will be set incorrectly in the output.